<html>
<head>
<title>Day 24: Exercise Solutions | Teclado</title>
<link href="../Styles/Style.css" type="text/css" rel="stylesheet"/>
</head>
<body>
<h1>第24天:锻炼解决方案| Teclado</h1>
<blockquote>原文：<a href="https://blog.teclado.com/python-30-day-24-exercise-solutions/">https://blog.teclado.com/python-30-day-24-exercise-solutions/</a></blockquote><div><article class="pb-12 prose prose-p:font-copy col-start-2 dark:prose-invert">
                    
                    <p>以下是我们对Python 系列<a href="https://blog.teclado.com/30-days-of-python/"> 30天</a><a href="/30-days-of-python/python-30-day-24-exceptions-advanced">第24天</a>练习的解决方案。在检查解决方案之前，请确保您亲自尝试了这些练习！</p>
<h3>1)要求用户输入一个介于1和10(包括1和10)之间的整数。如果他们给出的数字超出了指定的范围，则引发<code>ValueError</code>并通知他们他们的选择无效。</h3>
<p>首先，我们需要实际获取一些用户输入，并将其转换为整数。我们没有被告知要处理在这一步出现的任何异常，但是如果您愿意，您可以这样做。</p>
<div class="codehilite"><pre><span/><code><span class="n">number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">input</span><span class="p">(</span><span class="s2">"Please enter a number between 1 and 10: "</span><span class="p">))</span>
</code></pre></div>

<p>为了检查数量是否在指定的范围内，我们可以使用这样的条件:</p>
<div class="codehilite"><pre><span/><code><span class="n">number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">input</span><span class="p">(</span><span class="s2">"Please enter a number between 1 and 10: "</span><span class="p">))</span>

<span class="k">if</span> <span class="n">number</span> <span class="o">&gt;</span> <span class="mi">10</span> <span class="ow">or</span> <span class="n">number</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
    <span class="k">pass</span>
</code></pre></div>

<p>或者，我们可以定义一个范围，并查看该数字是否在该范围内:</p>
<div class="codehilite"><pre><span/><code><span class="n">number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">input</span><span class="p">(</span><span class="s2">"Please enter a number between 1 and 10: "</span><span class="p">))</span>

<span class="k">if</span> <span class="n">number</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">11</span><span class="p">):</span>
    <span class="k">pass</span>
</code></pre></div>

<p>在这两种情况下，我们都希望在条件语句体中引发一个<code>ValueError</code>,如下所示:</p>
<div class="codehilite"><pre><span/><code><span class="n">number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">input</span><span class="p">(</span><span class="s2">"Please enter a number between 1 and 10: "</span><span class="p">))</span>

<span class="k">if</span> <span class="n">number</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">11</span><span class="p">):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The number must be between 1 and 10. You entered </span><span class="si">{</span><span class="n">number</span><span class="si">}</span><span class="s2">."</span><span class="p">)</span>
</code></pre></div>

<h3>2)下面你会发现一个<code>divide</code>函数。编写异常处理，这样我们就可以捕获<code>ZeroDivisionError</code>异常、<code>TypeError</code>异常和其他种类的<code>ArithmeticError</code>。</h3>
<p>我们得到的函数是这样的:</p>
<div class="codehilite"><pre><span/><code><span class="n">divide</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">a</span> <span class="o">/</span> <span class="n">b</span><span class="p">)</span>
</code></pre></div>

<p>首先，我们需要处理任何<code>ZeroDivisionError</code>异常，我们可以这样做:</p>
<div class="codehilite"><pre><span/><code><span class="n">divide</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">a</span> <span class="o">/</span> <span class="n">b</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ZeroDivisionError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">"Cannot divide by zero"</span><span class="p">)</span>
</code></pre></div>

<p>接下来，我们需要捕捉更一般的<code>TypeError</code>异常，这可以通过添加另一个<code>except</code>子句来实现。</p>
<div class="codehilite"><pre><span/><code><span class="n">divide</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">a</span> <span class="o">/</span> <span class="n">b</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ZeroDivisionError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">"Cannot divide by zero"</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">"Both values must be numbers. Cannot divide </span><span class="si">{a}</span><span class="s2"> and </span><span class="si">{b}</span><span class="s2">"</span><span class="p">)</span>
</code></pre></div>

<p>最后，我们需要捕捉一般的<code>ArithmeticError</code>异常。</p>
<p>这最后的另一个条款很重要。确保不要把它放在<code>ZeroDivisionError</code>分支之前，因为<code>ZeroDivisionError</code>的所有实例都是<code>ArithmeticError</code>的实例。记住Python将按顺序一次检查一个<code>except</code>子句，并且它将使用它首先找到的匹配分支。</p>
<div class="codehilite"><pre><span/><code><span class="n">divide</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">a</span> <span class="o">/</span> <span class="n">b</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ZeroDivisionError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">"Cannot divide by zero"</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">"Both values must be numbers. Cannot divide </span><span class="si">{a}</span><span class="s2"> and </span><span class="si">{b}</span><span class="s2">"</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ArithmeticError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">"Could not complete the division. The numbers were likely too large"</span><span class="p">)</span>
</code></pre></div>

<h3>3)下面你会发现一个接受集合的<code>itemgetter</code>函数，以及一个键或索引。捕捉任何<code>KeyError</code>或<code>IndexError</code>的实例，并将异常写入一个名为<code>log.txt</code>的文件，同时附上导致这个问题的参数。写入日志文件后，重新引发原始异常。</h3>
<div class="codehilite"><pre><span/><code><span class="k">def</span> <span class="nf">itemgetter</span><span class="p">(</span><span class="n">collection</span><span class="p">,</span> <span class="n">identifier</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">collection</span><span class="p">[</span><span class="n">identifier</span><span class="p">]</span>
</code></pre></div>

<p>首先，我将编写另一个名为<code>log_exception</code>的函数，它将负责写入文件。这将使我们的异常处理更加简洁，因为我们不会在<code>except</code>子句中嵌入上下文管理器，等等。</p>
<div class="codehilite"><pre><span/><code><span class="k">def</span> <span class="nf">log_exception</span><span class="p">(</span><span class="n">exception</span><span class="p">,</span> <span class="n">fn</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="n">values</span> <span class="o">=</span> <span class="s2">", "</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">=</span><span class="si">{</span><span class="n">value</span><span class="si">!r}</span><span class="s2">"</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">"log.txt"</span><span class="p">,</span> <span class="s2">"a"</span><span class="p">)</span> <span class="k">as</span> <span class="n">log_file</span><span class="p">:</span>
        <span class="n">log_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Exception: </span><span class="si">{</span><span class="n">exception</span><span class="si">}</span><span class="s2">, Function: </span><span class="si">{</span><span class="n">fn</span><span class="si">}</span><span class="s2">, Values: </span><span class="si">{</span><span class="n">values</span><span class="si">}</span><span class="se">\n</span><span class="s2">"</span><span class="p">)</span>
</code></pre></div>

<p>这里的一切我们以前都见过，但让我们快速浏览一遍。</p>
<p>我们的函数有三个参数:<code>exception</code>、<code>fn</code>和<code>**kwargs</code>。<code>exception</code>将接收一个异常对象，我们将通过命名我们捕获的特定异常来获取该异常对象。<code>fn</code>将是一个代表函数名的字符串，这样我们就可以确定异常发生在哪里。<code>**kwargs</code>将收集传递给该函数的所有参数，以及赋值给这些参数的值。</p>
<p>在函数体的第一行，我们遍历了<code>**kwargs</code>中的值，并生成了一个字符串，将键和值表示为关键字参数。对于<code>value</code>,我们有这个特殊的<code>!r</code>,这是一种确保我们得到的表示与代码中值的出现方式相匹配的方式，而不是当我们打印它们时用户看到的样子。这意味着我们可以在日志中区分<code>4</code>和<code>"4"</code>。</p>
<p>一旦我们有了值，我们就用逗号将它们连接在一起，然后在append模式(<code>"a"</code>)下向日志文件中写入一行。</p>
<p>现在让我们来看看实际的异常处理。在这种情况下，我们希望对两个异常做同样的事情，所以我们有一些选择。我们要么抓住更宽泛的<code>LookupError</code>，要么同时抓住<code>IndexError</code>和<code>KeyError</code>。在这种情况下，我会选择后者。</p>
<div class="codehilite"><pre><span/><code><span class="k">def</span> <span class="nf">log_exception</span><span class="p">(</span><span class="n">exception</span><span class="p">,</span> <span class="n">fn</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="n">values</span> <span class="o">=</span> <span class="s2">", "</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">=</span><span class="si">{</span><span class="n">value</span><span class="si">!r}</span><span class="s2">"</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">"log.txt"</span><span class="p">,</span> <span class="s2">"a"</span><span class="p">)</span> <span class="k">as</span> <span class="n">log_file</span><span class="p">:</span>
        <span class="n">log_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Exception: </span><span class="si">{</span><span class="n">exception</span><span class="si">}</span><span class="s2">, Function: </span><span class="si">{</span><span class="n">fn</span><span class="si">}</span><span class="s2">, Values: </span><span class="si">{</span><span class="n">values</span><span class="si">}</span><span class="se">\n</span><span class="s2">"</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">itemgetter</span><span class="p">(</span><span class="n">collection</span><span class="p">,</span> <span class="n">identifier</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">collection</span><span class="p">[</span><span class="n">identifier</span><span class="p">]</span>
    <span class="k">except</span> <span class="p">(</span><span class="ne">IndexError</span><span class="p">,</span> <span class="ne">KeyError</span><span class="p">)</span> <span class="k">as</span> <span class="n">ex</span><span class="p">:</span>
        <span class="n">log_exception</span><span class="p">(</span><span class="n">ex</span><span class="p">,</span> <span class="s2">"itemgetter"</span><span class="p">,</span> <span class="n">collection</span><span class="o">=</span><span class="n">collection</span><span class="p">,</span> <span class="n">identifier</span><span class="o">=</span><span class="n">identifier</span><span class="p">)</span>
</code></pre></div>

<p>现在我们只需要使用<code>raise</code>关键字重新引发异常。</p>
<div class="codehilite"><pre><span/><code><span class="k">def</span> <span class="nf">log_exception</span><span class="p">(</span><span class="n">exception</span><span class="p">,</span> <span class="n">fn</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="n">values</span> <span class="o">=</span> <span class="s2">", "</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">=</span><span class="si">{</span><span class="n">value</span><span class="si">!r}</span><span class="s2">"</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">"log.txt"</span><span class="p">,</span> <span class="s2">"a"</span><span class="p">)</span> <span class="k">as</span> <span class="n">log_file</span><span class="p">:</span>
        <span class="n">log_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Exception: </span><span class="si">{</span><span class="n">exception</span><span class="si">}</span><span class="s2">, Function: </span><span class="si">{</span><span class="n">fn</span><span class="si">}</span><span class="s2">, Values: </span><span class="si">{</span><span class="n">values</span><span class="si">}</span><span class="se">\n</span><span class="s2">"</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">itemgetter</span><span class="p">(</span><span class="n">collection</span><span class="p">,</span> <span class="n">identifier</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">collection</span><span class="p">[</span><span class="n">identifier</span><span class="p">]</span>
    <span class="k">except</span> <span class="p">(</span><span class="ne">IndexError</span><span class="p">,</span> <span class="ne">KeyError</span><span class="p">)</span> <span class="k">as</span> <span class="n">ex</span><span class="p">:</span>
        <span class="n">log_exception</span><span class="p">(</span><span class="n">ex</span><span class="p">,</span> <span class="s2">"itemgetter"</span><span class="p">,</span> <span class="n">collection</span><span class="o">=</span><span class="n">collection</span><span class="p">,</span> <span class="n">identifier</span><span class="o">=</span><span class="n">identifier</span><span class="p">)</span>
        <span class="k">raise</span>
</code></pre></div>

<p>我们可以做得更好的一件事是处理<code>TypeError</code>异常，因为当我们将一个非整数作为一个序列的索引传入时会发生这种情况。</p>
<div class="codehilite"><pre><span/><code><span class="k">def</span> <span class="nf">log_exception</span><span class="p">(</span><span class="n">exception</span><span class="p">,</span> <span class="n">fn</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="n">values</span> <span class="o">=</span> <span class="s2">", "</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">=</span><span class="si">{</span><span class="n">value</span><span class="si">!r}</span><span class="s2">"</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">"log.txt"</span><span class="p">,</span> <span class="s2">"a"</span><span class="p">)</span> <span class="k">as</span> <span class="n">log_file</span><span class="p">:</span>
        <span class="n">log_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Exception: </span><span class="si">{</span><span class="n">exception</span><span class="si">}</span><span class="s2">, Function: </span><span class="si">{</span><span class="n">fn</span><span class="si">}</span><span class="s2">, Values: </span><span class="si">{</span><span class="n">values</span><span class="si">}</span><span class="se">\n</span><span class="s2">"</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">itemgetter</span><span class="p">(</span><span class="n">collection</span><span class="p">,</span> <span class="n">identifier</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">collection</span><span class="p">[</span><span class="n">identifier</span><span class="p">]</span>
    <span class="k">except</span> <span class="p">(</span><span class="ne">IndexError</span><span class="p">,</span> <span class="ne">KeyError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">)</span> <span class="k">as</span> <span class="n">ex</span><span class="p">:</span>
        <span class="n">log_exception</span><span class="p">(</span><span class="n">ex</span><span class="p">,</span> <span class="s2">"itemgetter"</span><span class="p">,</span> <span class="n">collection</span><span class="o">=</span><span class="n">collection</span><span class="p">,</span> <span class="n">identifier</span><span class="o">=</span><span class="n">identifier</span><span class="p">)</span>
        <span class="k">raise</span>
</code></pre></div>

<h3>注意</h3>
<p>我们实际上并不像我在这个练习中展示的那样进行日志记录，因为我们不需要这样做。Python有一个<code>logging</code>模块，它为我们做了很多工作，并生成更多信息日志。</p>
<p>你可以在这里找到关于日志记录的信息和入门教程。</p> 
                </article>
            
            </div>    
</body>
</html>